cmd/manager/main.go

```diff
@@ -105,100 +137,115 @@ func main() {
+		// Populate the global default OperatorSources definition and config
+		if err := defaults.PopulateGlobals(); err != nil {
+			logger.Fatal(err)
+		}
 
-	// Handle the defaults
-	err = ensureDefaults(cfg, mgr.GetScheme())
-	if err != nil {
-		logrus.Fatal(err)
-	}
+		logger.Info("setting up controllers")
+		if err := controller.AddToManager(mgr, options.ControllerOptions{}); err != nil {
+			logger.Fatal(err)
+		}
 
-	// statusReportingDoneCh will be closed after the operator has successfully stopped reporting ClusterOperator status.
-	statusReportingDoneCh := statusReporter.StartReporting()
+		logger.Info("ensuring the default catalogsource resources")
+		if err := ensureDefaults(cfg, mgr.GetScheme()); err != nil {
+			logger.Fatalf("failed to setup the default catalogsource manifests: %v", err)
+		}
 
-	// Start the Cmd
-	err = mgr.Start(stopCh)
+		// start reporting the marketplace clusteroperator status reporting before
+		// starting the manager instance as mgr.Start is blocking
+		logger.Info("starting the marketplace clusteroperator status reporter")
+		statusReportingDoneCh := statusReporter.StartReporting()
 
-	// Wait for ClusterOperator status reporting routine to close the statusReportingDoneCh channel.
-	<-statusReportingDoneCh
+		logger.Info("starting manager")
+		if err := mgr.Start(stopCh); err != nil {
+			logger.WithError(err).Error("unable to run manager")
+		}
 
-	exit(err)
-}
+		// Wait for ClusterOperator status reporting routine to close the statusReportingDoneCh channel.
+		<-statusReportingDoneCh
+	}
 
-// exit stops the reporting of ClusterOperator status and exits with the proper exit code.
-func exit(err error) {
-	// If an error exists then exit with status set to 1
+	client, err := kubernetes.NewForConfig(mgr.GetConfig())
 	if err != nil {
-		logrus.Fatalf("The operator encountered an error, exit code 1: %v", err)
+		logger.Fatal(fmt.Errorf("failed to initialize the kubernetes clientset: %v", err))
+	}
+
+	id := os.Getenv("POD_NAME")
+	if id == "" {
+		logger.Warn("failed to determine $POD_NAME falling back to hostname")
+		id, err = os.Hostname()
+		if err != nil {
+			logger.Fatal(err)
+		}
 	}
 
-	// No error, graceful termination
-	logrus.Info("The operator exited gracefully, exit code 0")
-	os.Exit(0)
+	rl := &resourcelock.ConfigMapLock{
+		Client: client.CoreV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity: id,
+		},
+		ConfigMapMeta: v1.ObjectMeta{
+			Name:      defaultLeaderElectionConfigMapName,
+			Namespace: leaderElectionNamespace,
+		},
+	}
+	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+		Lock:            rl,
+		ReleaseOnCancel: true,
+		LeaseDuration:   defaultLeaseDuration,
+		RenewDeadline:   defaultRenewDeadline,
+		RetryPeriod:     defaultRetryPeriod,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				logger.Infof("became leader: %s", id)
+				run(leaderCtx)
+			},
+			OnStoppedLeading: func() {
```

Comment:
@timflannagan could you explain what it means for this controller to give up leadership? Should it stop managing CatalogSources?

If so, I'm having trouble finding where that's happening.

Reply:
Both the old and new leader election implementations are centered around a ConfigMap locking mechanism. In the previous implementation, which used operator-sdk's leader-for-life mechanism, was stalling upgrades entirely when the node holding the lock being unreachable, and the result is that a new marketplace-operator pod is unable to preempt the previous pod as the ConfigMap leader election lease cannot be garbage collected.

Now, preemption is handled using the leader-for-lease implementation, where a pod/node will obtain a lock, become a leader, and periodically will attempt to renew this lease. In the case of multiple pods all attempting to obtain that lease and become a leader, one of those pods will obtain leadership, and in the next preemption poll, the current leader will need to attempt to re-acquire leadership and if that process is unsuccessful, another will be elected leader and lock ownership will transition to that pod's metadata.Name (i.e. their identity).